### ISSUE TITLE
Data augmentation

### ISSUE THREAD

**Issue by bhack** [keras-team/keras-tuner#153]
Do you plan to support/document/colab Keras tuner use in augmentation pipelines?

https://blog.insightdatascience.com/automl-for-data-augmentation-e87cf692c366

https://arxiv.org/abs/1905.07373

**Comment by bhack** [comment]
As reference I add also https://github.com/tensorflow/models/blob/master/research/autoaugment/README.md

**Comment by omalleyt12** [comment]
@bhack Thanks for the issue!

This can be done via overriding `Tuner.run_trial` with custom preprocessing / augmentation. Hyperparameters can be added anywhere inside `run_trial`. Here's a simple example:

```python
class AugmentTuner(kt.Tuner):
  def run_trial(self, trial, x, y, ...):
    hp = trial.hyperparameters
    x = horizontal_flip(x, prob=hp.Float('flip_prob', 0, 1, step=0.1))
    super(AugmentTuner, self).run_trial(trail, x, y, ...)
```

Here's a more in-depth tutorial on subclassing Tuner: https://keras-team.github.io/keras-tuner/tutorials/subclass-tuner/

**Comment by bhack** [comment]
Thanks, but I think that a full e2e example or colab could be useful cause in real case probably the data will use also TF.data pipelines.
Do you think that Tuner.run_trial Is enough also for implementing OHL-Auto-Aug like strategies (Algorithm 1 in the 1st paper)?

**Comment by omalleyt12** [comment]
@bhack The tutorial linked uses `tf.data`. In general, you can use the same methods you would use in a regular `tf.data` pipeline, simply replace concrete parameters with hyperparameters as necessary. 

Online hyperparameter tuning as described in the linked paper (i.e. changing hyperparameter values within a single training run) would be pretty difficult to do with this framework, although it's doable. You would have to write your own `Oracle` and formulate each iteration with different hyperparameter values (the outer training loop in the paper) as its own trial

Hope that helps!

**Comment by bhack** [comment]
Yes thanks. If you are going to close this ticket can we place some hint about data auto augmentation somewhere? 
Just to be sure to have a referenze about this use case of the tuner for other interested users/developers.

**Comment by omalleyt12** [comment]
@bhack sure we can mention it in the Tuner subclassing tutorial

**Comment by bhack** [comment]
Is there any current Relationship between keras-tuner and autokeras repo under keras-team?

**Comment by omalleyt12** [comment]
@bhack yep, autokeras relies on keras-tuner

**Comment by bhack** [comment]
Mhh I see many augmentation stuff there with just a quick code search but I don't know if it is something specific to "smart" auto augmentation policies like the ones mentioned early in this issue.

**Comment by Uiuran** [comment]
There is no recipe to Data Augmentation in AutoML yet, it is subject of research. Also, it is questionable how much Data Augmentation can improve accuracy and not overfit. Seems to be useful to have the plug/pipe anyway, but to be part of the algorithm still subject of research ...

**Comment by bhack** [comment]
IMHO also AutoML Is subject of research in Deep learning. :smile: 
So I think It Is a legitimate topic.
I've attached this topic to a more general topic in the AutoML repo https://github.com/keras-team/autokeras/issues/828

**Comment by omalleyt12** [comment]
Hyperparameter tuning is definitely helpful for data augmentation. Closing this issue out as this is possible via overriding `Tuner.run_trial`, please continue the discussion on AutoKeras as to what should be included "out of the box" in that library (I think there is some plan for doing data augmentation in AutoKeras, but I'm not clear on the specifics)

**Comment by bhack** [comment]
Check https://github.com/keras-team/autokeras/issues/828#issuecomment-555542084
